### ISSUE TITLE
How to debug program terminating with uncaught exception of type PAL_SEHException

### ISSUE THREAD

**Issue by liiri** [dotnet/runtime#13685]
I'm testing how my dotnet core application behaves on no internet connection. The program always crashes with

```
libc++abi.dylib: terminating with uncaught exception of type PAL_SEHException
```
I tried to find what command is triggering it but could not find it by manual trial and error.

I know it has something to do with the network connection since I don't get this crash when network is stable. I'm seeing fine responses from my external API calls (failing with WebException).

How can I debug this?

I'm using dotnet core `3.0.100` and `OSX 10.14.5`

**Comment by eltiare** [comment]
I'm getting this under Rider, too. It might be something with their debugger.

**Comment by Parasrah** [comment]
`terminate called after throwing an instance of 'PAL_SEHException'`

I'm getting this using VSCode on PopOS 19.10 as well

**Comment by xWiiLLz** [comment]
Just FYI, got that same error while debugging a dotnet core project inside a Docker container using VSCode's debugger: `terminate called after throwing an instance of 'PAL_SEHException'
watch : Exited with error code 134`.

Dotnet project running inside Docker (`Docker version 18.09.5, build e8ff056`), host running `Ubuntu 18.04.2 LTS` (bionic) & accessing debugger (`CoreCLR` packaged with `ms-vscode.csharp`) through a VSCode remote SSH to the Ubuntu Host.

**Comment by tobbe303** [comment]
Having the same issue debugging from VSCode on MacOS 10.15.3 and dotnet core 3.1.102.

**Comment by mehyaa** [comment]
Same error on debug:

```
Mac OS X 10.15.3
JetBrains Rider 2019.3.4
dotnet 3.1.102
```

```
Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
terminating with uncaught exception of type PAL_SEHException
abort() called
```

**Comment by liiri** [comment]
@janvorli do you think issue should be reopened? I haven't managed to get a stack trace, but maybe someone else can

**Comment by tobbe303** [comment]
Here's some info including the stack from the thread that crashed.

```
Process:               dotnet [12869]
Path:                  /usr/local/share/dotnet/dotnet
Identifier:            dotnet
Version:               0
Code Type:             X86-64 (Native)
Parent Process:        vsdbg-ui [12868]
Responsible:           Terminal [624]
User ID:               501

Date/Time:             2020-03-05 09:36:11.727 +0100
OS Version:            Mac OS X 10.15.3 (19D76)
Report Version:        12
Bridge OS Version:     4.2 (17P3050)
Anonymous UUID:        74D1C03A-2DB9-1A21-29D4-49CC2D13727A

Sleep/Wake UUID:       CA123E96-9314-43A4-93D6-F26E4FE3DBC0

Time Awake Since Boot: 21000 seconds
Time Since Wake:       6700 seconds

System Integrity Protection: enabled

Crashed Thread:        29

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
terminating with uncaught exception of type PAL_SEHException
abort() called
```

```
Thread 29 Crashed:
0   libsystem_kernel.dylib        	0x00007fff63f447fa __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff64001bc1 pthread_kill + 432
2   libsystem_c.dylib             	0x00007fff63ecba1c abort + 120
3   libc++abi.dylib               	0x00007fff60f65be8 abort_message + 231
4   libc++abi.dylib               	0x00007fff60f65d9c demangling_terminate_handler() + 262
5   libobjc.A.dylib               	0x00007fff62a91792 _objc_terminate() + 104
6   libc++abi.dylib               	0x00007fff60f72dc7 std::__terminate(void (*)()) + 8
7   libc++abi.dylib               	0x00007fff60f6520f __cxa_rethrow + 99
8   libcoreclr.dylib              	0x00000001062ec043 RaiseTheExceptionInternalOnly(Object*, int, int) + 1251
9   libcoreclr.dylib              	0x00000001064b3277 Thread::HandleThreadAbort() + 583
10  libcoreclr.dylib              	0x0000000106438caa JIT_NewArr1(CORINFO_CLASS_STRUCT_*, long) + 122
11  ???                           	0x000000010cc788be 0 + 4509370558
12  libcoreclr.dylib              	0x0000000106576c6b CallDescrWorkerInternal + 124
13  libcoreclr.dylib              	0x00000001063c28e9 MethodDescCallSite::CallTargetWorker(unsigned long const*, unsigned long*, int) + 1657
14  libcoreclr.dylib              	0x0000000106263a97 FuncEvalWrapper(MethodDescCallSite*, DebuggerEval*, unsigned long const*, unsigned char*) + 151
15  libcoreclr.dylib              	0x00000001062660cd DoNormalFuncEval(DebuggerEval*, unsigned char*, Object**, void**, void**, long*, ValueClassInfo**) + 1597
16  libcoreclr.dylib              	0x0000000106264c57 GCProtectArgsAndDoNormalFuncEval(DebuggerEval*, unsigned char*) + 2023
17  libcoreclr.dylib              	0x00000001062654ff FuncEvalHijackWorker + 335
18  libcoreclr.dylib              	0x0000000106272a52 FuncEvalHijack + 14
19  ???                           	0x0000000110297ffb 0 + 4566122491
20  ???                           	0x0000000110424a7c 0 + 4567747196
21  ???                           	0x0000000110451e8d 0 + 4567932557
22  ???                           	0x00000001104245a1 0 + 4567745953
23  ???                           	0x0000000110422990 0 + 4567738768
24  ???                           	0x000000010f05dcbd 0 + 4547009725
25  ???                           	0x00000001104530ae 0 + 4567937198
26  ???                           	0x000000010f05ead6 0 + 4547013334
27  ???                           	0x000000010f05dea0 0 + 4547010208
28  ???                           	0x000000011045329c 0 + 4567937692
29  ???                           	0x0000000110458b64 0 + 4567960420
30  ???                           	0x0000000110425024 0 + 4567748644
31  ???                           	0x000000010f05dcbd 0 + 4547009725
32  ???                           	0x000000011045408e 0 + 4567941262
33  ???                           	0x000000010f05ead6 0 + 4547013334
34  ???                           	0x000000010f05dea0 0 + 4547010208
35  ???                           	0x000000011045329c 0 + 4567937692
36  ???                           	0x0000000110458b64 0 + 4567960420
37  ???                           	0x000000011042577b 0 + 4567750523
38  ???                           	0x000000010f05dcbd 0 + 4547009725
39  ???                           	0x000000011045408e 0 + 4567941262
40  ???                           	0x000000010f05ead6 0 + 4547013334
41  ???                           	0x000000010f05dea0 0 + 4547010208
42  ???                           	0x000000011045329c 0 + 4567937692
43  ???                           	0x0000000110458b64 0 + 4567960420
44  ???                           	0x000000011042698e 0 + 4567755150
45  ???                           	0x000000010f05dcbd 0 + 4547009725
46  ???                           	0x000000011045408e 0 + 4567941262
47  ???                           	0x000000010f05ead6 0 + 4547013334
48  ???                           	0x000000010f05dea0 0 + 4547010208
49  ???                           	0x000000011045329c 0 + 4567937692
50  ???                           	0x0000000110458b64 0 + 4567960420
51  ???                           	0x0000000110444757 0 + 4567877463
52  ???                           	0x000000010f05dcbd 0 + 4547009725
53  ???                           	0x000000011045408e 0 + 4567941262
54  ???                           	0x000000010f05ead6 0 + 4547013334
55  ???                           	0x000000010f05dea0 0 + 4547010208
56  ???                           	0x000000011045329c 0 + 4567937692
57  ???                           	0x0000000110458b64 0 + 4567960420
58  ???                           	0x0000000110444cc1 0 + 4567878849
59  ???                           	0x000000010f05dcbd 0 + 4547009725
60  ???                           	0x000000011045408e 0 + 4567941262
61  ???                           	0x000000010f05ead6 0 + 4547013334
62  ???                           	0x000000010f05dea0 0 + 4547010208
63  ???                           	0x000000011045329c 0 + 4567937692
64  ???                           	0x0000000110458b64 0 + 4567960420
65  ???                           	0x0000000110445234 0 + 4567880244
66  ???                           	0x000000010f05dcbd 0 + 4547009725
67  ???                           	0x000000011045408e 0 + 4567941262
68  ???                           	0x000000010f05ead6 0 + 4547013334
69  ???                           	0x000000010f05dea0 0 + 4547010208
70  ???                           	0x000000010f064d08 0 + 4547038472
71  ???                           	0x000000010f064c33 0 + 4547038259
72  ???                           	0x000000011042c324 0 + 4567778084
73  ???                           	0x000000010f05dcbd 0 + 4547009725
74  ???                           	0x000000011045a09e 0 + 4567965854
75  ???                           	0x000000010f05ead6 0 + 4547013334
76  ???                           	0x000000010f05dea0 0 + 4547010208
77  ???                           	0x000000010f064d08 0 + 4547038472
78  ???                           	0x000000011045ac87 0 + 4567968903
79  ???                           	0x000000011045acc9 0 + 4567968969
80  ???                           	0x0000000110440386 0 + 4567860102
81  ???                           	0x000000011044001a 0 + 4567859226
82  ???                           	0x000000011043f971 0 + 4567857521
83  ???                           	0x000000011043f751 0 + 4567856977
84  ???                           	0x000000011043f5da 0 + 4567856602
85  ???                           	0x000000011043f571 0 + 4567856497
86  ???                           	0x000000010cb19d1a 0 + 4507933978
87  libcoreclr.dylib              	0x0000000106576c6b CallDescrWorkerInternal + 124
88  libcoreclr.dylib              	0x00000001063c28e9 MethodDescCallSite::CallTargetWorker(unsigned long const*, unsigned long*, int) + 1657
89  libcoreclr.dylib              	0x00000001063dd7d8 QueueUserWorkItemManagedCallback(void*) + 184
90  libcoreclr.dylib              	0x0000000106383883 ManagedThreadBase_DispatchOuter(ManagedThreadCallState*) + 323
91  libcoreclr.dylib              	0x0000000106383ea0 ManagedThreadBase::ThreadPool(void (*)(void*), void*) + 32
92  libcoreclr.dylib              	0x000000010637a4d4 ManagedPerAppDomainTPCount::DispatchWorkItem(bool*, bool*) + 276
93  libcoreclr.dylib              	0x00000001063a7e8f ThreadpoolMgr::WorkerThreadStart(void*) + 1311
94  libcoreclr.dylib              	0x00000001061fbad4 CorUnix::CPalThread::ThreadEntry(void*) + 436
95  libsystem_pthread.dylib       	0x00007fff64001e65 _pthread_start + 148
96  libsystem_pthread.dylib       	0x00007fff63ffd83b thread_start + 15
```

**Comment by janvorli** [comment]
@tobbe303 thank you a lot for providing the call stack. Now it is clear what's causing the trouble. It is a ThreadAbort exception stemming from FuncEval that the debugger calls to evaluate managed properties. This exception is expected to happen when there is a Property that has code that takes too long to execute and debugger decides to stop it as it could be a genuine hang. It needs to be investigated why we don't handle it properly on Linux. 
I can quickly see one difference already which seems to be likely the culprit - the `FuncEvalHijack` assembler helper uses `UnhandledExceptionHandlerUnix` personality routine on Unix while it uses `FuncEvalHijackPersonalityRoutine` on Windows. We don't have an implementation of that method for Unix. 

cc: @dotnet/dotnet-diag - I will look into implementing the `FuncEvalHijackPersonalityRoutine` for Unix

**Comment by janvorli** [comment]
Actually, there is one more thing and the personality routine is probably not the culprit since I've realized that the exception should not pass through the FuncEvalHijack frame.

The real culprit seems to be that `Thread::HandleThreadAbort` throws the `PAL_SEHException` for the ThreadAbort exception and it should be caught in JIT_NewArr1 as it uses the HELPER_METHOD_FRAME_BEGIN_RET_0 and propagated further using the managed exception handling through the frame 11 etc . But it seems that for some reason, it doesn't get caught there, the C++ unwinder hits the managed frame and since it doesn't know anything about it, it considers the exception unhandled and aborts the process.

**Comment by titanix** [comment]
I got this error too, after charging a big XML file. Both with VS Code debugger and by running directly `dotnet run`.

**Comment by eltiare** [comment]
I suppose I should note that I'm getting this occasionally on MacOS.

**Comment by janvorli** [comment]
@tobbe303 can you please tell me which version of .NET Core was running when you got this stack trace?

**Comment by janvorli** [comment]
I have actually found the real culprit. Here is the problem. Managed code executed as part of the func eval ends up calling `JIT_NewArr1` native function. In our case, when that function is entered, the debugger has already marked the thread for aborting the execution because the func eval was taking too long. The `JIT_NewArr1` uses  the `HELPER_METHOD_FRAME_BEGIN_RET_0` macro to create a HelperMethodFrame explicit frame. But this macro also does few additional things. Together with the `HELPER_METHOD_FRAME_END` macro, it creates a try / catch around the body of the method that allows native exceptions to be caught and then processed by the DispatchManagedException, as we cannot let it propagate to the managed code.
Besides that, it also calls `HelperMethodFrame::Push()` to push the explicit frame on the per thread explicit frame list. That function further calls (tail-calls in release build) `HelperMethodFrame::PushSlowHelper()` which checks if the thread is marked for abortion and if it is, it calls (tail-calls in release build) `HelperMethodFrame::HandleThreadAbort`. And that method calls `RaiseTheExceptionInternalOnly` which throws `PAL_SEHException` to initiate the thread abort propagation.
The problem is that this is done before the try / catch region. So the `PAL_SEHException` is not caught and turned into calling the `DispatchManagedException` as expected, but gets propagated into the managed caller. And since the libunwind doesn't know anything about the managed code, the exception is considered unhandled and C++ runtime aborts the process.

It seems the fix might be as easy as moving the try / catch (`INSTALL_MANAGED_EXCEPTION_DISPATCHER` / `UNINSTALL_MANAGED_EXCEPTION_DISPATCHER` macros) so that the try region contains the `HelperMethodFrame::Push()`, but it is possible that there will be some devil hidden in the details.
